typescript/src/schema/proto.ts

```diff
@@ -17,10 +38,114 @@ interface MethodDescriptorProto
   inputInterface: string;
   outputInterface: string;
   comments: string;
+  methodConfig: plugin.grpc.service_config.MethodConfig;
+  retryableCodesName: string;
+  retryParamsName: string;
+  timeoutMillis?: number;
+}
+
+class RetryableCodeMap {
+  codeEnumMapping: { [index: string]: string };
+  uniqueCodesNamesMap: { [uniqueName: string]: string };
+  prettyCodesNamesMap: { [prettyName: string]: string[] };
+  uniqueParamsNamesMap: { [uniqueName: string]: string };
+  prettyParamNamesMap: { [prettyName: string]: {} };
+
+  constructor() {
+    this.uniqueCodesNamesMap = {};
+    this.prettyCodesNamesMap = {};
+    this.uniqueParamsNamesMap = {};
+    this.prettyParamNamesMap = {};
+
+    // build reverse mapping for enum: 0 => OK, 1 => CANCELLED, etc.
+    this.codeEnumMapping = {};
+    const allCodes = Object.keys(plugin.google.rpc.Code);
+    for (const code of allCodes) {
+      this.codeEnumMapping[
+        ((plugin.google.rpc.Code as unknown) as {
+          [key: string]: plugin.google.rpc.Code;
+        })[code].toString()
+      ] = code;
+    }
+
+    // generate some pre-defined code sets for compatibility with existing configs
+    this.getRetryableCodesName(
+      defaultNonIdempotentCodes,
+      defaultNonIdempotentRetryCodesName
+    );
+    this.getRetryableCodesName(
+      defaultIdempotentCodes,
+      defaultIdempotentRetryCodesName
+    );
+    this.getParamsName(defaultParameters, 'default');
+  }
+
+  private buildUniqueCodesName(
+    retryableStatusCodes: plugin.google.rpc.Code[]
+  ): string {
+    // generate an unique readable name for the given retryable set of codes
+    const sortedCodes = retryableStatusCodes.sort(
+      (a, b) => Number(a) - Number(b)
+    );
+    const uniqueName = sortedCodes
+      .map(code => this.codeEnumMapping[code])
+      .join('_')
+      .toSnakeCase();
+    return uniqueName;
+  }
+
+  private buildUniqueParamsName(params: {}): string {
+    // generate an unique not so readable name for the given set of parameters
+    return objectHash(params);
```

Comment:
I'm not super familiar with hashing, but if I have two retry policies that contain all of the same values, but for different MethodConfigs (maybe there is something else different about the MethodConfig other than the RetryPolicy), how do we know that this will in fact be unique?

Reply:
Here I only care about the contents of `RetryPolicy` so it should be fine.

Actually, I think the logic of this class is complicated enough to be covered by unit tests, not just by baseline tests. I will add some unit tests that will likely explain the use cases better than just the code :)

Reply:
SGTM, thanks